gap/oper.gi

```diff
@@ -765,6 +765,91 @@ function(D1, D2, edge_function)
   return Digraph(edges);
 end);
 
+InstallMethod(AmalgamDigraphsIsomorphic,
+"for a digraph, a digraph, a list, and a list",
+[IsDigraph, IsDigraph, IsList, IsList],
+function(D1, D2, subdigraphVertices1, subdigraphVertices2)
+  local subdigraph1, subdigraph2, newSubdigraphVertices2, transformation, vertex;
+
+  subdigraph1 := InducedSubdigraph(D1, subdigraphVertices1);
+  subdigraph2 := InducedSubdigraph(D2, subdigraphVertices2);
+
+  if not IsIsomorphicDigraph(subdigraph1, subdigraph2) then
+    ErrorNoReturn(
+      "the two subdigraphs must be isomorphic.");
+  fi;
+
+  newSubdigraphVertices2 := [];
+  transformation := DigraphEmbedding(subdigraph2, subdigraph1);
+  for vertex in subdigraphVertices2 do
+    newSubdigraphVertices2[
+      Position(subdigraphVertices2, vertex) ^ transformation] := vertex;
+  od;
+
+  return AmalgamDigraphs(D1, D2, subdigraphVertices1, newSubdigraphVertices2);
+end);
+
+InstallMethod(AmalgamDigraphs,
+"for a digraph, a digraph, a list, and a list",
+[IsDigraph, IsDigraph, IsList, IsList],
+function(D1, D2, subdigraphVertices1, subdigraphVertices2)
+  local D, map, vertex, vertexList, size, iterator, edgeList, subLength;
+
+  if not InducedSubdigraph(D1, subdigraphVertices1) =
+         InducedSubdigraph(D2, subdigraphVertices2) then
+    ErrorNoReturn(
+      "the two subdigraphs must be equal.");
+  fi;
+
+  # Create a mutable copy so that the function also works on
+  # immutable input digraphs.
+  D := DigraphMutableCopy(D1);
+  subLength := Length(subdigraphVertices1);
+
+  # 'map' is a mapping from the vertices of D2 to the vertices of the
+  # final output graph. The idea is to map the subdigraph vertices of D2
+  # onto the subdigraph vertices of D1 and then map the rest of the vertices
+  # of D2 to other (higher) values. The mapping from D1 to the output graph
+  # can be understood as the identity mapping.
+  map := rec();
+
+  for vertex in [1 .. subLength] do
+    map.(subdigraphVertices2[vertex]) := subdigraphVertices1[vertex];
+  od;
+
+  vertexList := Difference(DigraphVertices(D2), subdigraphVertices2);
+  size := DigraphNrVertices(D1);
+  iterator := 1;
+  for vertex in vertexList do
+    map.(vertex) := iterator + size;
+    iterator := iterator + 1;
+  od;
+
+  # The problem with adding edges to the output graph was that the
+  # edges of of the subdigraph were added twice, creating multiple
+  # edges between certain pairs of points. A quick and readable fix
+  # would have been to use DigraphRemoveAllMultipleEdges, but I decided
+  # to check each of the edges being added to see if they were already
+  # in the subdigraph. This way the function does not end up adding edges
+  # only to delete them later.
+  edgeList := ShallowCopy(DigraphEdges(D2));
```

Comment:
I'd be tempted to just go with `DigraphRemoveAllMultipleEdges` here or maybe:

~~~
DigraphAddVertices(D, DigraphNrVertices(D2) - Length(subdigraphVertices2));

for edge in DigraphEdges(D2) do
  if not (edge[1] in subdigraphVertices2 and edge[2] in subdigraphVertices2) then
    DigraphAddEdge(D, edge);
  fi;
od
~~~

Reply:
Ok, so based on the code here and some of the comments above, I just wanted to double check quickly if I understood the overall idea of the function correctly.

Essentially, I can't seem to find a way to construct the map in such a way that it could be represented as a transformation and without using something like the lines 820-826 and I also noticed this piece of code seems to be placing the edges differently from the way I would have expected. So I'm wondering if I've understood everything right.

As far as I understand it, if I want to add vertices to D  in order to create the output graph, then I need to have a way to map the non-subdigraph vertices of D2 to the newly added vertices (which will have a higher index than the number of vertices in D1). This is sort of what I had in mind when constructing the 'map' and is also what I intended to do with the lines 820-826 above. It also means that the map variable can't be represented as a Transformation as far as I know, since there will often be entries in map that will be larger than the length of the map (since map needs to have the same number of entries as D2).